### PR TITLE
Assistants - chat bots with skills

### PR DESCRIPTION
**Pull request by cgillum**
This PR adds support for "assistants" - i.e., chat bots that can invoke functions as custom skills.

### What's new

To support assistants, a new `AssistantSkillTrigger` is introduced. Any such functions in your project become automatically callable by chat bots. A sample demonstrating this is also included in the PR.

### How it works

As part of this change, whenever a prompt is sent to a chat bot, the list of `AssistantSkillTrigger` functions are sent to the GPT language model. The model can then decide to respond by invoking one of those functions, which is done by the updated entity logic. The response is then sent back to the model so that it can respond to the user prompt appropriately.

image
*(Image is borrowed from a Japanese presentation, introducing this idea)*

### Remaining work

- [x] **An actual implementation**: The original prototype depends on the Durable Functions implementation of chat bots. However, that's being replaced as part of [this PR](https://github.com/Azure/azure-functions-openai-extension/pull/10). Once merged, we can add the assistant functionality to the updated chat bot implementation.
- [ ] **Support for more language stacks**: Currently this feature only works with .NET in-proc. More work is needed to get it to work with .NET Isolated, Python, TypeScript, etc.
- [ ] **Integration tests**: We have an integration test for chat bots, but there should also be one for assistants with skills.

### Future improvements

- **Support for complex parameter types**: Currently the skill triggers only support primitive types (string, int, float, bool, etc.).
* **Fine-grained access controls for skills**: Right now, all bots can call any skill defined in the function app. It might be worth considering whether certain skills can be restricted to specific assistants.
* **User confirmation**: Support marking certain skills as "requiring user confirmation" and then having the bot request permission before invoking that skill.

### CODE REVIEW

**Comment by cgillum** [comment]
A few other minor changes I made as part of this PR:

- Fixed several warnings in the existing code (nullability, etc.)
- Renamed some types to improve clarity
- Fixed inconsistencies in code style
- Refactored and simplified ChatBotService logic
- Added missing cancellation token flow to chat bot service logic
- Minor (breaking) schema changes to reduce data storage

File: samples/assistant/README.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+The sample is available in the following language stacks:
+
+* [C# on the in-process worker](csharp-inproc)
+
+You must have the following installed on your local machine in order to run these samples.
+
+* [.NET 6 SDK](https://dotnet.microsoft.com/download/dotnet/6.0) or newer
+* [Azure Functions Core Tools](https://learn.microsoft.com/azure/azure-functions/functions-run-local?tabs=v4%2Cwindows%2Ccsharp%2Cportal%2Cbash) v4.0.5455 or newer
+* An [OpenAI API key](https://platform.openai.com/account/api-keys) saved into a `OPENAI_API_KEY` environment variable
+* Azure Storage emulator such as [Azurite](https://learn.microsoft.com/azure/storage/common/storage-use-azurite) running in the background
```
Comment (manvkaur) [inline samples/assistant/README.md:49]: Root level read me can be referenced here.

As we also want to share the keys for Azure settings.

File: samples/assistant/README.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+If you want to run the sample with Cosmos DB, then you must also have the following installed:
+
+* [Azure Cosmos DB Emulator](https://docs.microsoft.com/azure/cosmos-db/local-emulator) or a connection string to a real Azure Cosmos DB resource
```
Comment (manvkaur) [inline samples/assistant/README.md:53]: also a mention to local.settings.json or app settings for Azure would be good.

File: samples/assistant/README.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
+# Assistants
```
Comment (manvkaur) [inline samples/assistant/README.md:1]: There should also be a reference to Assistant README in root level where all extension features are listed.

File: samples/assistant/README.md [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+
+The `AssistantSkillTrigger` attribute requires a `FunctionDescription` string value, which is text describing what the function does.
+This is critical for the AI assistant to be able to invoke the skill at the right time.
+The name of the function parameter (e.g., `taskDescription`) is also an important hint to the AI assistant about what kind of information to provide to the skill.
+
+> **NOTE:** The `AssistantSkillTrigger` attribute only supports primitive types as function parameters, such as `string`, `int`, `bool`, etc. Function return values can be of any JSON-serializable type.
+
+Any function that uses the `AssistantSkillTrigger` binding will be automatically registered as a skill that can be invoked by any AI assistant.
+The assistant will invoke a skill function whenever it decides to do so to satisfy a user prompt, and will provide function parameters based on
+the context of the conversation. The skill function can then return a response to the assistant, which will be used to satisfy the user prompt.
```
Comment (manvkaur) [inline samples/assistant/README.md:36]: nit: new line after the word `on` can be avoided

**Comment by manvkaur** [comment]
We can add the pending tasks or future improvements to our internal board backlog

File: src/WebJobs.Extensions.OpenAI/Models/ChatMessage.cs [hunk 5]
```diff
@@ -25,7 +27,7 @@ public ChatMessageEntity(string content, string role)
     public string Content { get; set; }
 
     /// <summary>
-    /// Gets or sets the role of the chat agent (“system”, “user”, "assistant", "function" or "tool").
+    /// Gets or sets the role of the chat agent.
     /// </summary>
     public string Role { get; set; }
```
[the remark this answers is not in the document]
Comment (cgillum) [reply src/WebJobs.Extensions.OpenAI/Models/ChatMessage.cs:32]: The problem I ran into when trying this was that `ChatRole` types don't serialize to JSON correctly. For example, it would serialize to `{}` instead of `"user"` or `"assistant"`. I decided to leave it as string so that serialization would work as expected.

File: samples/other/dotnet/csharp-inproc/Demos/demo.http [hunk 6]
```diff
@@ -0,0 +1,27 @@
+### Ingest first email into database
+
+POST http://localhost:7127/api/IngestEmail
+Content-Type: application/json
+
+{"FilePath":"C:\\Users\\cgillum\\Downloads\\email\\Re_ OpenAI Idea-A-Thon.txt"}
```
Comment (cgillum) [inline samples/other/dotnet/csharp-inproc/Demos/demo.http:6]: I didn't mean to include this file. I'll remove it from the PR.

File: samples/assistant/README.md (hunk 2 above)
Comment (cgillum) [reply samples/assistant/README.md:53]: Yes, will add this.

File: samples/assistant/README.md (hunk 1 above)
Comment (cgillum) [reply samples/assistant/README.md:49]: Ah, yes, will reference the root-level README.

File: samples/assistant/README.md (hunk 4 above)
Comment (cgillum) [reply samples/assistant/README.md:36]: I added this because the line was getting too long to read in an editor without word-wrap, but I can remove it.

File: samples/assistant/README.md [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+AI assistants are chat bots that can be configured with custom skills. Custom skills are implemented using the `assistantSkillTrigger` binding.
+Custom skills are useful ways to extend the functionality of an AI assistant. For example, you can create a custom skill that saves a todo item to a database, or a custom skill that queries a database for a list of todo items.
+In both cases, the skills are defined in your app and any the language model doesn't need to know anything about the skill implementation, making it useful for interacting with private data.
+
+This OpenAI extension internally uses the [function calling](https://platform.openai.com/docs/guides/function-calling) functionality available in
+`gpt-3.5-turbo` and `gpt-4` models to implement assistant skills.
+
+## Defining skills
+
+You can define a skill by creating a function that uses the `AssistantSkillTrigger` binding. The following C# in-proc example shows a skill that adds a todo item to a database:
```
Comment (aishwaryabh) [inline samples/assistant/README.md:17]: Is it possible to make this work with OOP in this PR or would we want to do this in another PR?

File: samples/assistant/README.md [hunk 8]
```diff
@@ -0,0 +1,221 @@
+# Assistants
+
+This sample demonstrates how to build an AI assistant chatbot with custom skills using Azure Functions and a local build of the experimental OpenAI extension.
```
[the remark this answers is not in the document]
Comment (cgillum) [reply samples/assistant/README.md:3]: Good catch, I'll just remove the word "experimental". Readers should know from context (hopefully) that everything in this repo is preview and I don't want to worry about forgetting to update this again later. :)

File: samples/assistant/README.md (hunk 3 above)
Comment (cgillum) [reply samples/assistant/README.md:1]: Ah, yes, I forgot to include that. Will add.

File: src/Functions.Worker.Extensions.OpenAI/ChatMessage.cs [hunk 9]
```diff
@@ -25,7 +27,7 @@ public ChatMessageEntity(string content, string role)
     public string Content { get; set; }
 
     /// <summary>
-    /// Gets or sets the role of the chat agent (“system”, “user”, "assistant", "function" or "tool").
+    /// Gets or sets the role of the chat agent.
     /// </summary>
-    public string Role { get; set; }
+    public ChatRole Role { get; set; }
```
Comment (aishwaryabh) [inline src/Functions.Worker.Extensions.OpenAI/ChatMessage.cs:32]: Does this work with the OOP chat bot scenario? I think it should just seralize but just wanted to validate :)

**Comment by cgillum** [comment]
BTW, here's what the conversation history looks like in Azure Table Storage after running the demo script.

image

File: src/Functions.Worker.Extensions.OpenAI/ChatMessage.cs (hunk 9 above)
Comment (cgillum) [reply src/Functions.Worker.Extensions.OpenAI/ChatMessage.cs:32]: Good question, let me check that. I might need to revert this change...

File: samples/assistant/README.md (hunk 7 above)
Comment (cgillum) [reply samples/assistant/README.md:17]: Unfortunately, more design work will be required for all out-of-proc languages since we don't have any way to automatically describe the function signatures in those languages. Let's create a separate work item for supporting out-of-process languages, including .NET Isolated.

File: src/Functions.Worker.Extensions.OpenAI/ChatMessage.cs (hunk 9 above)
Comment (aishwaryabh) [reply src/Functions.Worker.Extensions.OpenAI/ChatMessage.cs]: Thanks for checking! I already have the work item on my plate. I'll start working on it as soon as we publish the nuget packages

**Review by aishwaryabh — APPROVED** [review]
LGTM!

File: CHANGELOG.md [hunk 10]
```diff
@@ -5,6 +5,12 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
```
Comment (manvkaur) [inline CHANGELOG.md:8]: @aishwaryabh , when you make a PR for Worker csproj update, please add the version add or change logs here

**Review by manvkaur — APPROVED** [review]
